src/OrleansRuntime/Core/InsideRuntimeClient.cs

```diff
@@ -414,6 +427,36 @@ public void SniffIncomingMessage(Message message)
             }
         }
 
+        private void RegisterCancellationToken(IAddressable target, object arg, InvokeMethodRequest request, int i)
+        {
+            var grainToken = ((GrainCancellationToken)arg);
+            if (grainToken.WentThroughSerialization)
+            {
+                CancellationSourcesExtension cancellationExtension;
+                if (!SiloProviderRuntime.Instance.TryGetExtensionHandler(out cancellationExtension))
+                {
+                    cancellationExtension = new CancellationSourcesExtension();
+
+                    if (!SiloProviderRuntime.Instance.TryAddExtension(cancellationExtension))
+                    {
+                        logger.Error(
+                            ErrorCode.CancellationExtensionCreationFailed,
+                            string.Format("Could not add cancellation token extension, target: {0}", target));
+                        return;
+                    }
+                }
+
+                request.Arguments[i] = cancellationExtension.GetOrCreateCancellationToken(
+                    grainToken.Id,
+                    grainToken.CancellationToken.IsCancellationRequested);
+            }
+            else
+            { 
+                // no need to register remote token extension for local token
+                request.Arguments[i] = grainToken;
```

Comment:
In `GrainCancellationToken` that came from deserialization unlinked `CancellationToken` is kept. This lines exists in order to provide new  `CancellationToken` that can be cancelled through `CancellationTokenSource` which located in the grain extension.